test/parser.go

```diff
@@ -0,0 +1,109 @@
+package main
+
+import (
+	"bufio"
+	"io"
+	"strings"
+	"errors"
+)
+
+// An Action is one instance of a START/END tag pair. ActionType is
+// the word following START. Commands are the lines between the tags.
+type Action struct {
+	ActionType string
+	Commands   []string
+	closed     bool
+}
+
+type stateFn func(*bufio.Scanner, *[]Action) stateFn
+
+func neutralState(scanner *bufio.Scanner, actions *[]Action) stateFn {
+	if scanner.Scan() == false {
+		return nil
+	}
+	line := scanner.Text()
+	words := strings.Fields(line)
+	for index, word := range words {
+		if word == "END" {
+			// Shouldn't find an END in neutral state.
+			return nil
+		}
+		if word == "START" {
+			action := words[index+1]
+			newAction(action, actions)
+			return insideActionState
+		}
+	}
+	return neutralState
+}
+
+func newAction(aType string, actions *[]Action) {
+	aa := Action{
+		ActionType: aType,
+		Commands:   make([]string, 0),
+		closed:     false,
+	}
+	*actions = append(*actions, aa)
+}
+
+func actionLine(line string, actions []Action) {
+	actions[len(actions)-1].Commands =
+		append(actions[len(actions)-1].Commands, line)
+}
+
+func closeAction(actions []Action) {
+	actions[len(actions)-1].closed = true
+}
+
+func insideActionState(scanner *bufio.Scanner, actions *[]Action) stateFn {
+	if scanner.Scan() == false {
+		return nil
+	}
+	line := scanner.Text()
+	words := strings.Fields(line)
+	if strings.HasPrefix(line, "```") ||  words[0] == "```" {
+		// Ignore line.
+		return insideActionState
+	}
+	for _, word := range words {
+		if word == "END" {
+			closeAction(*actions)
+			return neutralState
+		}
+		if word == "START" {
+			// Shouldn't find a START inside an action.
+			return nil
+		}
+	}
+	actionLine(line, *actions)
+	return insideActionState
+}
+
+
+// GetActions parses the input for START/END tags and compiles a list
+// of actions, comprising of the lines between the lines with
+// START/END. Lines that begin with ``` are ignored. The word
+// immeidately after START is saved as the action type.
+func GetActions(reader io.Reader) ([]Action, error) {
+	scanner := bufio.NewScanner(reader)
+	actions := make([]Action, 0)
+
+	for state := neutralState; state != nil; {
+		state = state(scanner, &actions)
```

Comment:
yeah, after thinking a little more, I think it's actually pretty good. I think you can move the handling of `scanner` into the `GetActions` function, and then all you need to pass to each state function is the line.